# Exploring eBay Car Sales Data

In this project, we'll work with a dataset of used cars from [*eBay Kleinanzeigen*](https://www.ebay-kleinanzeigen.de/), a classifieds section of the German eBay website. 

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/), and we will be working with 50,000 data points from the full dataset. The original dataset is no longer available on Kaggle, bit it can be found [here](https://data.world/data-society/used-cars-data).

The data dictionary has columns labeled as follows:

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `odometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the listings by brand, price, and mileage.

## Import Data

First, we will import the data for cleaning. As `UTF-8` encoding did not work, we have encoded with `Latin-1`.

In [1]:
import pandas as pd
import numpy as np
import datetime as dt
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

A first look at the dataset shows that of the 20 columns, there are 15 columns with values stored as strings and five columns with values stored as integers. There are only five columns with any null values, and every column has greater than 80% non-null values. Finally, we see that the columns are listed in [camel case](https://en.wikipedia.org/wiki/Camel_case).

In order to make the data a bit easier to use, we'll change the name of some of the columns and convert the names to [snake case](https://en.wikipedia.org/wiki/Snake_case), the general convention when using Python.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
new_columns = []

for c in autos.columns:
    if c == 'yearOfRegistration':
        c = 'registration_year'
    elif c =='monthOfRegistration':
        c = 'registration_month'
    elif c == 'notRepairedDamage':
        c = 'unrepaired_damage'
    elif c =='dateCreated':
        c = 'ad_created'
    else:
        c = c.lower()
    new_columns.append(c)

autos.columns = new_columns
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Here is a look at the dataset with the column name changes that we made.

In [6]:
autos.describe(include='all')

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

## Data Cleaning

There is quite a bit of interesting information that we can decipher from this look at `autos`. We can make quite a few changes to the data in order to simplify it and make our analysis easier.

1. `nrofpictures` has a value of `0` (zero) for every row and can be removed.
2. `datecrawled`, `adcreated`, and `lastseen` can be convered from strings to datetime objects.
3. `price` and `odometer` are text values and should be converted to numbers before further examination.
4. `registration_year` and `registration_month` have bad data as there are values that fall outside of the reasonable ranges for these columns.
5. `seller` and `offertype` have only one instance where the value is different.
6. Where possible, translate information from German into English.

Let's take each of these one-by-one.

### 1. Remove the `nrofpictures` Column

This column has a value of `0` (zero) for every row and therefore provides no valuable information. Let's remove it to speed up our data processing.

In [7]:
autos = autos.drop('nrofpictures', axis=1)

### 2. Convert the `datecrawled`, `adcreated`, and `lastseen` Columns to Datetime Objects

The `datecrawled`, `adcreated`, and `lastseen` columns are all string-type objects but contain dates and times. While we can work with these columns as strings without issue, conversion will make our code more intuitive. Best practice should always be to keep our data in the most logical format. This helps us better understand what we were doing when we return to the code, and it also makes it easier for others to continue our work and/or make changes.

In [8]:
autos['datecrawled'] = pd.to_datetime(autos['datecrawled'])
autos['ad_created'] = pd.to_datetime(autos['ad_created'])
autos['lastseen'] = pd.to_datetime(autos['lastseen'])

print(type(autos['datecrawled'][3]))
print(type(autos['ad_created'][27]))
print(type(autos['lastseen'][45]))

print('\n')
print(autos['ad_created'].describe())

<class 'pandas._libs.tslibs.timestamps.Timestamp'>
<class 'pandas._libs.tslibs.timestamps.Timestamp'>
<class 'pandas._libs.tslibs.timestamps.Timestamp'>


count                   50000
unique                     76
top       2016-04-03 00:00:00
freq                     1946
first     2015-06-11 00:00:00
last      2016-04-07 00:00:00
Name: ad_created, dtype: object


This is a quick and simple change, and we can see that we have done this properly. Let's continue with the next item on our list.

### 3. Convert the `price` and `odometer` Columns to Numbers

The `price` and `odometer` columns contain numeric values, but we can see that these two columns are strings instead of numbers because of the addition of different characters. Let's look at the unique values of these columns to see what characters need to be removed before we convert them.

In [9]:
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [10]:
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

We need to remove `"$"` and `","` from `autos` and `","` and `"km"` from `odometer`. Since the odometer readings are in kilometers (km), let's rename the `odometer` column as `odometer_km` for clarity.

In [11]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','')
autos['price'] = autos['price'].astype(int)

In [12]:
autos['odometer'] = autos['odometer'].str.replace(',','')
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].astype(int)

In [13]:
autos.rename({'odometer': 'odometer_km'}, axis=1)

datecrawled                                               name  \
0     2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1     2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2     2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3     2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4     2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                   ...                                                ...   
49995 2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996 2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997 2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998 2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999 2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offertype  price   abtest vehicletype  registration_year  \
0      privat   Angebot   5000  control         bus               2004   
1      privat   Angebot   8500  control   limousine               1997   
2      privat   Angebot   8990     test   limousine               2009   
3      privat   Angebot   4350  control  kleinwagen               2007   
4      privat   Angebot   1350     test       kombi               2003   
...       ...       ...    ...      ...         ...                ...   
49995  privat   Angebot  24900  control   limousine               2011   
49996  privat   Angebot   1980  control      cabrio               1996   
49997  privat   Angebot  13200     test      cabrio               2014   
49998  privat   Angebot  22900  control       kombi               2013   
49999  privat   Angebot   1250  control   limousine               1996   

         gearbox  powerps   model  odometer_km  registration_month fueltype  \
0        manuell      158  andere       150000                   3      lpg   
1      automatik      286     7er       150000                   6   benzin   
2        manuell      102    golf        70000                   7   benzin   
3      automatik       71  fortwo        70000                   6   benzin   
4        manuell        0   focus       150000                   7   benzin   
...          ...      ...     ...          ...                 ...      ...   
49995  automatik      239      q5       100000                   1   diesel   
49996    manuell       75   astra       150000                   5   benzin   
49997  automatik       69     500         5000                  11   benzin   
49998    manuell      150      a3        40000                  11   diesel   
49999    manuell      101  vectra       150000                   1   benzin   

            brand unrepaired_damage ad_created  postalcode            lastseen  
0         peugeot              nein 2016-03-26       79588 2016-04-06 06:45:54  
1             bmw              nein 2016-04-04       71034 2016-04-06 14:45:08  
2      volkswagen              nein 2016-03-26       35394 2016-04-06 20:15:37  
3           smart              nein 2016-03-12       33729 2016-03-15 03:16:28  
4            ford              nein 2016-04-01       39218 2016-04-01 14:38:50  
...           ...               ...        ...         ...                 ...  
49995        audi              nein 2016-03-27       82131 2016-04-01 13:47:40  
49996        opel              nein 2016-03-28       44807 2016-04-02 14:18:02  
49997        fiat              nein 2016-04-02       73430 2016-04-04 11:47:27  
49998        audi              nein 2016-03-08       35683 2016-04-05 16:45:07  
49999        opel              nein 2016-03-13       45897 2016-04-06 21:18:48  

[50000 rows x 19 columns]

We can see above that there are only 13 different values in the `odometer_km` column, so they were likely input as a range. Regardless, that data looks sufficient, so let's look at the `price` column to see if there is data that doesn't look right and needs to be changed or removed.

In [14]:
autos['price'].unique().shape

(2357,)

In [15]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [16]:
autos['price'].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
            ... 
5              2
3              1
2              3
1            156
0           1421
Name: price, Length: 2357, dtype: int64

For the `price` column, we can see that there are 2,357 unique prices, but there are a number of prices that are very high with eight prices listed as greater than or equal to 10,000,000. There are also 1,421 listings where the price is 0 (zero). Since this data is scraped from eBay, it is possible that there are listings where the starting bid is below one dollar, resulting in data that round to zero. Let's look in more detail at the most expensive listings.

In [17]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

There are some extremely high prices here and while all of these prices look very high for a used car, there is a relatively steady increase in price up to `$350,000` before the price jumps up to one million dollars and beyond. Let's remove every listing where the price is above `$350,000`

In [18]:
autos = autos[autos['price'] <= 350000]

### 4. Fix Bad Data in `registration_year` and `registration_month` Columns

When we looked at these columns earlier, we could see that there was incorrect data in both of these columns. The `registration_year` column has a minimum value of 1000, long before cars were invented, and a maximum value of 9999, a date which obviously we have not yet reached. The `registration_month` column contains no values greater than 12, but does contain at least one row with a value of 0 (zero).

Both of these columns are already of type integer, so we do not need to make any changes there. Every value that falls outside of a reasonable range, we will convert to `NaN` (not a number) so that we can easily keep those rows separate from the ones that contain data with values we can more readily accept as accurate.

Let's begin by exploring the `registration_year` column in a bit more detail.

In [19]:
autos['registration_year'].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2005, 2002, 2012, 2011,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1967, 1973, 1956, 1976, 4500, 1987, 1991, 1983, 1960, 1969,
       1950, 1978, 1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971,
       1966, 1979, 1981, 1970, 1974, 1910, 1975, 5000, 4100, 2019, 1959,
       9996, 9999, 6200, 1964, 1958, 1800, 1948, 1931, 1943, 9000, 1941,
       1962, 1927, 1937, 1929, 1000, 1957, 1952, 1111, 1955, 1939, 8888,
       1954, 1938, 2800, 5911, 1500, 1953, 1951, 4800, 1001])

In [20]:
autos['registration_year'].value_counts().sort_index(ascending=False).head(20)

9999       4
9996       1
9000       2
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       3
2018     491
2017    1452
2016    1316
2015     399
2014     665
2013     806
2012    1323
2011    1634
Name: registration_year, dtype: int64

In [21]:
autos['registration_year'].value_counts().sort_index(ascending=False).tail(15)

1943    1
1941    2
1939    1
1938    1
1937    4
1934    2
1931    1
1929    1
1927    1
1910    9
1800    2
1500    1
1111    1
1001    1
1000    1
Name: registration_year, dtype: int64

In [22]:
x = (autos['registration_year'].between(1910, 2016)).sum() / autos.shape[0] * 100
print('\033[1m' + 'Number of vehicles registered between 1910 and 2016: '
      + '\033[0m' + str(x.round(2)) + "%")

Number of vehicles registered between 1910 and 2016: 96.06%


We can see that on both ends of the scale there are only a few listings with dates that are obviously inaccurate. However, as this data is from 2016, it is theoretically impossible for a car to have a registration year greater than 2016. 

However, given the distribution of vehicles according to the `registration_year` column, it is likely that these dates do not correspond to the date of most recent registration but the original registration date or model year of the vehicle. Since the model years for cars release early and most 2017 vehicles are actually released in 2016, we should include all of the listings for 2017 as well as these could be vehicle pre-orders or models that were released very early in the year.

In [23]:
autos = autos.loc[autos['registration_year'].between(1910, 2017)]

Now, let's look at the `registration_month` column

In [24]:
autos['registration_month'].value_counts()

3     5029
0     4951
6     4320
5     4066
4     4060
7     3921
10    3625
12    3411
9     3363
11    3331
1     3242
8     3169
2     2980
Name: registration_month, dtype: int64

While `0` (zero) is not a valid registration month, we can see that it is the second-most common value for our dataset. Since the registration year values suggest that many of the registration dates may be the model year instead of an actual date, `0` (zero) would be a logical value where the month of release is unknown. As such, we'll make no further revisions to the data based on the `registration_month` column.

### 5. Remove the `seller` and `offertype` column

These columns have only one row where the value is different and can likely be removed, but let's look at each one before doing so.

In [25]:
autos['seller'].value_counts()

privat        49467
gewerblich        1
Name: seller, dtype: int64

There are 49,467 instances of `privat` (German: private) sales and one row listed as `gewerblich` (German: commercial). While we could leave the data as it is, excluding non-private sales will possibly allow us to draw better conclusions while only necessitating the removal of one row. Let's remove this row and the `seller` column.

In [26]:
autos = autos.loc[autos['seller'] == 'privat']

autos = autos.drop('seller', axis=1)

In [27]:
autos['offertype'].value_counts()

Angebot    49466
Gesuch         1
Name: offertype, dtype: int64

Again, there is only one value which is different. There are 49,466 instances of `Angebot` (German: offer) and one that is 'Gesuch' (German: wanted). We are looking at car sales data, so a listing where someone is looking to buy a specific car is not relevant to our analysis. Let's remove this row and the `offertype` column.

In [28]:
autos = autos.loc[autos['offertype'] == 'Angebot']

autos = autos.drop('offertype', axis=1)

### 6. Translate Information from German into English

Since this data was taken from a German website, there is a fair bit of information that is in German. Where feasible, let's translate this information into English.

In [29]:
autos['gearbox'].unique()

array(['manuell', 'automatik', nan], dtype=object)

These values correspond to manual (German: "manuell") and automatic (German: "automatik") transmissions as well as some listings where the type is not listed. Let's translate this column into English.

In [30]:
autos.loc[autos['gearbox'] == 'manuell', 'gearbox'] = 'manual'
autos.loc[autos['gearbox'] == 'automatik', 'gearbox'] = 'automatic'

In [31]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

This is a simple yes (German: "ja") or no (German: "nein") question. We'll translate these values accordingly.

In [32]:
autos.loc[autos['unrepaired_damage'] == 'ja', 'unrepaired_damage'] = 'yes'
autos.loc[autos['unrepaired_damage'] == 'nein', 'unrepaired_damage'] = 'no'

In [33]:
autos['fueltype'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

There are a number of fuel types, but the only ones that we need to change are "benzin" to gasoline, "elektro" to electric, and "andere" to other.

In [34]:
autos.loc[autos['fueltype'] == 'benzin', 'fueltype'] = 'gasoline'
autos.loc[autos['fueltype'] == 'elektro', 'fueltype'] = 'electric'
autos.loc[autos['fueltype'] == 'andere', 'fueltype'] = 'other'

In [35]:
autos['vehicletype'].value_counts()

limousine     12854
kleinwagen    10822
kombi          9124
bus            4091
cabrio         3061
coupe          2533
suv            1986
andere          420
Name: vehicletype, dtype: int64

The final column we'll translate is the `vehicletype` column.  While, the fact that the most common type of vehicle listed is limousine, a quick look at some of the vehicles listed as this type are certainly not limousines. While a further examination of the datset in this area could yield some interesting results, that is beyond the scope of this analysis. We'll simply translate the words to English for coherency and leave the rest of the analysis to a different study.

In [36]:
autos.loc[autos['vehicletype'] == 'kleinwagen', 'vehicletype'] = 'compact'
autos.loc[autos['vehicletype'] == 'kombi', 'vehicletype'] = 'station_wagon'
autos.loc[autos['vehicletype'] == 'cabrio', 'vehicletype'] = 'convertible'
autos.loc[autos['vehicletype'] == 'andere', 'vehicletype'] = 'other'

Now that the majority of the cleaning is complete, we'll look at the dataset one last time before moving on to analysis.

In [37]:
autos

datecrawled                                               name  \
0     2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1     2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2     2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3     2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4     2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                   ...                                                ...   
49995 2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996 2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997 2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998 2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999 2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price   abtest    vehicletype  registration_year    gearbox  powerps  \
0       5000  control            bus               2004     manual      158   
1       8500  control      limousine               1997  automatic      286   
2       8990     test      limousine               2009     manual      102   
3       4350  control        compact               2007  automatic       71   
4       1350     test  station_wagon               2003     manual        0   
...      ...      ...            ...                ...        ...      ...   
49995  24900  control      limousine               2011  automatic      239   
49996   1980  control    convertible               1996     manual       75   
49997  13200     test    convertible               2014  automatic       69   
49998  22900  control  station_wagon               2013     manual      150   
49999   1250  control      limousine               1996     manual      101   

        model  odometer  registration_month  fueltype       brand  \
0      andere    150000                   3       lpg     peugeot   
1         7er    150000                   6  gasoline         bmw   
2        golf     70000                   7  gasoline  volkswagen   
3      fortwo     70000                   6  gasoline       smart   
4       focus    150000                   7  gasoline        ford   
...       ...       ...                 ...       ...         ...   
49995      q5    100000                   1    diesel        audi   
49996   astra    150000                   5  gasoline        opel   
49997     500      5000                  11  gasoline        fiat   
49998      a3     40000                  11    diesel        audi   
49999  vectra    150000                   1  gasoline        opel   

      unrepaired_damage ad_created  postalcode            lastseen  
0                    no 2016-03-26       79588 2016-04-06 06:45:54  
1                    no 2016-04-04       71034 2016-04-06 14:45:08  
2                    no 2016-03-26       35394 2016-04-06 20:15:37  
3                    no 2016-03-12       33729 2016-03-15 03:16:28  
4                    no 2016-04-01       39218 2016-04-01 14:38:50  
...                 ...        ...         ...                 ...  
49995                no 2016-03-27       82131 2016-04-01 13:47:40  
49996                no 2016-03-28       44807 2016-04-02 14:18:02  
49997                no 2016-04-02       73430 2016-04-04 11:47:27  
49998                no 2016-03-08       35683 2016-04-05 16:45:07  
49999                no 2016-03-13       45897 2016-04-06 21:18:48  

[49466 rows x 17 columns]

One last correction that can be made is to remove the `abtest` column. [A/B testing](https://en.wikipedia.org/wiki/A/B_testing) is a form of data analysis for websites. While this column might be useful to someone looking at metrics for eBay Germany, it does not provide any valuable information to us, so we will remove the column.

In [38]:
autos = autos.drop('abtest', axis=1)

In [39]:
autos

datecrawled                                               name  \
0     2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1     2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2     2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3     2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4     2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                   ...                                                ...   
49995 2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996 2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997 2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998 2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999 2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price    vehicletype  registration_year    gearbox  powerps   model  \
0       5000            bus               2004     manual      158  andere   
1       8500      limousine               1997  automatic      286     7er   
2       8990      limousine               2009     manual      102    golf   
3       4350        compact               2007  automatic       71  fortwo   
4       1350  station_wagon               2003     manual        0   focus   
...      ...            ...                ...        ...      ...     ...   
49995  24900      limousine               2011  automatic      239      q5   
49996   1980    convertible               1996     manual       75   astra   
49997  13200    convertible               2014  automatic       69     500   
49998  22900  station_wagon               2013     manual      150      a3   
49999   1250      limousine               1996     manual      101  vectra   

       odometer  registration_month  fueltype       brand unrepaired_damage  \
0        150000                   3       lpg     peugeot                no   
1        150000                   6  gasoline         bmw                no   
2         70000                   7  gasoline  volkswagen                no   
3         70000                   6  gasoline       smart                no   
4        150000                   7  gasoline        ford                no   
...         ...                 ...       ...         ...               ...   
49995    100000                   1    diesel        audi                no   
49996    150000                   5  gasoline        opel                no   
49997      5000                  11  gasoline        fiat                no   
49998     40000                  11    diesel        audi                no   
49999    150000                   1  gasoline        opel                no   

      ad_created  postalcode            lastseen  
0     2016-03-26       79588 2016-04-06 06:45:54  
1     2016-04-04       71034 2016-04-06 14:45:08  
2     2016-03-26       35394 2016-04-06 20:15:37  
3     2016-03-12       33729 2016-03-15 03:16:28  
4     2016-04-01       39218 2016-04-01 14:38:50  
...          ...         ...                 ...  
49995 2016-03-27       82131 2016-04-01 13:47:40  
49996 2016-03-28       44807 2016-04-02 14:18:02  
49997 2016-04-02       73430 2016-04-04 11:47:27  
49998 2016-03-08       35683 2016-04-05 16:45:07  
49999 2016-03-13       45897 2016-04-06 21:18:48  

[49466 rows x 16 columns]

## Dataset Analysis

There is a great deal of information in this dataset, but we will limit our analysis to brand, price, and mileage.

### Listing by Brand

In [40]:
autos['brand'].value_counts()

volkswagen        10562
opel               5390
bmw                5382
mercedes_benz      4689
audi               4254
ford               3449
renault            2367
peugeot            1449
fiat               1287
seat                915
skoda               782
mazda               751
nissan              749
smart               695
citroen             692
toyota              615
sonstige_autos      531
hyundai             485
volvo               454
mini                424
mitsubishi          399
honda               391
kia                 353
alfa_romeo          327
porsche             293
suzuki              289
chevrolet           279
chrysler            181
dacia               128
daihatsu            126
jeep                109
subaru              107
land_rover           99
daewoo               78
saab                 78
trabant              77
jaguar               76
rover                69
lancia               56
lada                 29
Name: brand, dtype: int64

In [41]:
autos['brand'].value_counts().shape

(40,)

In [42]:
autos['brand'].value_counts().head(1) / autos.shape[0]

volkswagen    0.21352
Name: brand, dtype: float64

In [43]:
autos['brand'].value_counts().head(10).sum() / autos.shape[0]

0.8034609630857559

### Analyzing Top Brands by Price

Examining the listings by brand, there are 40 different brands of vehicle listed for sale, but the top handful of brands make up a large percentage of the total listings. On its own, Volkswagen has nearly twice as many listings as any other brand and makes up more than 21% of all offers. We'll explore pricing for the top ten brands as these ten make up more than 80% of all listings.

In [44]:
brand_labels = autos['brand'].value_counts().head(10)
print(brand_labels)

volkswagen       10562
opel              5390
bmw               5382
mercedes_benz     4689
audi              4254
ford              3449
renault           2367
peugeot           1449
fiat              1287
seat               915
Name: brand, dtype: int64


In [45]:
avg_price_by_brand = {}


for b in brand_labels.index:
    selected_rows = autos[autos['brand'] == b]
    mean = round(selected_rows['price'].mean(), 2)
    avg_price_by_brand[b] = mean
    
print('\033[1m' + "Average Price of Vehicle by Brand, Top Ten Brands" + '\033[0m')
print('\n')  
for p in avg_price_by_brand:
    print('\033[1m' + p + ": " +'\033[0m' + str(avg_price_by_brand[p]))
    

Average Price of Vehicle by Brand, Top Ten Brands


volkswagen: 5178.13
opel: 2853.82
bmw: 8051.63
mercedes_benz: 8384.58
audi: 8989.55
ford: 3632.72
renault: 2365.32
peugeot: 3014.34
fiat: 2693.5
seat: 4259.55


Of the top 10 brands:

* Audi, Mercedes Benz, and BMW cost the most by a significant margin, nearly double all other brands in the top ten.
* Renault, Fiat, Opel are the cheapest brands with all three having an average cost less than one-third the average cost of the most expensive brand, Audi.
* Volkswagen, the most popular listing is very much in between these ranges, costing significantly less than the top three brands but significantly more than the rest of the top ten, more than twice the cost of the cheapest brand, Renault.

### Analyzing Top Brands by Mileage

With the large discrepancy in average price by brand among the top ten brands, it is important to look at the average mileage for each vehicle to see if this plays a factor in price.

In [46]:
avg_odometer_by_brand = {}


for b in brand_labels.index:
    selected_rows = autos[autos['brand'] == b]
    mean = round(selected_rows['odometer'].mean(), 2)
    avg_odometer_by_brand[b] = mean
    
print('\033[1m' + "Average Odometer of Vehicle by Brand, Top Ten Brands" + '\033[0m')
print('\n')  
for o in avg_odometer_by_brand:
    print('\033[1m' + o + ": " +'\033[0m' + str(avg_odometer_by_brand[o]))
    

Average Odometer of Vehicle by Brand, Top Ten Brands


volkswagen: 128968.47
opel: 129322.82
bmw: 132517.65
mercedes_benz: 131037.53
audi: 129561.59
ford: 124079.44
renault: 128221.38
peugeot: 127311.94
fiat: 116810.41
seat: 122027.32


In [47]:
total = 0

for o in avg_odometer_by_brand:
    total += avg_odometer_by_brand[o]
    
print(round(total / 10))

126986.0


In [48]:
apb_series = pd.Series(avg_price_by_brand)
aob_series = pd.Series(avg_odometer_by_brand)

df = pd.DataFrame(apb_series, columns=['mean_price'])
new_df = df.assign(mileage_in_km = aob_series)

new_df

mean_price  mileage_in_km
volkswagen        5178.13      128968.47
opel              2853.82      129322.82
bmw               8051.63      132517.65
mercedes_benz     8384.58      131037.53
audi              8989.55      129561.59
ford              3632.72      124079.44
renault           2365.32      128221.38
peugeot           3014.34      127311.94
fiat              2693.50      116810.41
seat              4259.55      122027.32

Looking at the mileage, this does not seem to be a factor in pricing. The only brand with an average mileage more than 5% from the mean mileage is Fiat, one of the cheapest brands. In fact, the top four brands in price also have the top four values for average mileage although not in the same order. This means that we can return to our original observations on prices of the top brands.

## Conclusion

Of the top 10 brands:

* Audi, Mercedes Benz, and BMW cost the most by a significant margin, nearly double all other brands in the top ten.
* Renault, Fiat, Opel are the cheapest brands with all three having an average cost less than one-third the average cost of the most expensive brand, Audi.
* Volkswagen, the most popular listing is very much in between these ranges, costing significantly less than the top three brands but significantly more than the rest of the top ten, more than twice the cost of the cheapest brand, Renault.
* Mileage is not a factor in vehicle cost by brand.

